cmd/keymasterd/unseal.go

```diff
@@ -0,0 +1,178 @@
+package main
+
+import (
+	"bytes"
+	"encoding/base64"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"io/ioutil"
+	"net/http"
+	"time"
+
+	"golang.org/x/crypto/openpgp"
+	"golang.org/x/crypto/openpgp/armor"
+
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/aws/arn"
+	"github.com/aws/aws-sdk-go/aws/credentials"
+	"github.com/aws/aws-sdk-go/aws/credentials/ec2rolecreds"
+	"github.com/aws/aws-sdk-go/aws/ec2metadata"
+	"github.com/aws/aws-sdk-go/aws/session"
+	"github.com/aws/aws-sdk-go/service/secretsmanager"
+)
+
+func (state *RuntimeState) secretInjectorHandler(w http.ResponseWriter,
+	r *http.Request) {
+	// checks this is only allowed when using TLS client certs.. all other authn
+	// mechanisms are considered invalid... for now no authz mechanisms are in
+	// place i.e. Any user with a valid cert can use this handler
+	if r.TLS == nil {
+		state.writeFailureResponse(w, r, http.StatusInternalServerError, "")
+		logger.Printf("We require TLS\n")
+		return
+	}
+	if len(r.TLS.VerifiedChains) < 1 {
+		state.writeFailureResponse(w, r, http.StatusForbidden, "")
+		logger.Printf("Forbidden\n")
+		return
+	}
+	clientName := r.TLS.VerifiedChains[0][0].Subject.CommonName
+	logger.Printf("Got connection from %s", clientName)
+	r.ParseForm()
+	sshCAPassword, ok := r.Form["ssh_ca_password"]
+	if !ok {
+		state.writeFailureResponse(w, r, http.StatusBadRequest,
+			"Invalid Post, missing data")
+		logger.Printf("missing ssh_ca_password")
+		return
+	}
+	if err := state.unsealCA([]byte(sshCAPassword[0]), clientName); err != nil {
+		state.writeFailureResponse(w, r, http.StatusBadRequest,
+			"Invalid Post, "+err.Error())
+		logger.Println(err)
+		return
+	}
+	w.WriteHeader(200)
+	fmt.Fprintf(w, "OK\n")
+	//fmt.Fprintf(w, "%+v\n", r.TLS)
+}
+
+func (state *RuntimeState) tryAutoUnseal() error {
+	baseConfig := state.Config.Base
+	if baseConfig.UnsealPasswordSecretArn == "" {
+		return nil
+	}
+	secretArn, err := arn.Parse(baseConfig.UnsealPasswordSecretArn)
+	if err != nil {
+		return err
+	}
+	if baseConfig.UnsealPasswordSecretKey == "" {
```

Comment:
This should be poart of the config loading.. we dont modify configs on the consumer

Reply:
It's not changing the config, only a local copy.

Reply:
As discussed offline, I've split out the setting of the default.